src/lib.rs

```diff
@@ -1,528 +1,564 @@
-            .round() as u16,
+    pub fn textures(&mut self) -> &mut Textures<Texture> {
+        &mut self.textures
+    }
+
+    pub fn atlas(&mut self) -> &Texture {
+        &mut self.textures.get(self.atlas).unwrap()
+    }
+
+    pub fn render<'a>(
+        &mut self,
+        ui: Ui<'a>,
+        device: &mut wgpu::Device,
+        encoder: &mut wgpu::CommandEncoder,
+        view: &wgpu::TextureView,
+    ) -> RendererResult<()> {
+        let FrameSize {
+            logical_size: (width, height),
+            hidpi_factor,
+        } = ui.frame_size();
+
+        if !(width > 0.0 && height > 0.0) {
+            return Ok(());
+        }
+        let fb_size = (
+            (width * hidpi_factor) as f32,
+            (height * hidpi_factor) as f32,
         );
 
-        rpass.draw_indexed(start..end, base_vertex as i32, 0..1);
-          
-        start = end;
-      }
-      Ok(())
-  }
-
-  fn update_uniform_buffer(
-    &mut self,
-    matrix: &[[f32; 4]; 4],
-  ) -> RendererResult<()> {
-    self.uniform_buffer.set_sub_data(0, cast_slice(matrix));
-    Ok(())
-  }
-
-  fn upload_vertex_buffer(
-    &mut self,
-    _device: &mut wgpu::Device,
-    vtx_buffer: &[ImDrawVert],
-//  ) -> RendererResult<wgpu::Buffer> {
-  ) -> RendererResult<()> {
-    let vertex_count = vtx_buffer.len() as u32;
-    if self.vertex_count + vertex_count < self.vertex_max {
-      self.vertex_buffer.set_sub_data(self.vertex_count * (size_of::<ImDrawVert>() as u32), cast_slice(vtx_buffer));
-      self.vertex_count += vertex_count;
-      Ok(())
+        let matrix = [
+            [(2.0 / width) as f32, 0.0, 0.0, 0.0],
+            [0.0, (2.0 / height) as f32, 0.0, 0.0],
+            [0.0, 0.0, -1.0, 0.0],
+            [-1.0, -1.0, 0.0, 1.0],
+        ];
+
+        let mut rpass = encoder.begin_render_pass(&wgpu::RenderPassDescriptor {
+            color_attachments: &[wgpu::RenderPassColorAttachmentDescriptor {
+                attachment: &view,
+                resolve_target: None,
+                load_op: match self.clear_color {
+                    Some(_) => wgpu::LoadOp::Clear,
+                    _ => wgpu::LoadOp::Load,
+                },
+                store_op: wgpu::StoreOp::Store,
+                clear_color: self.clear_color.unwrap_or(wgpu::Color {
+                    r: 0.0,
+                    g: 0.0,
+                    b: 0.0,
+                    a: 1.0,
+                }),
+            }],
+            depth_stencil_attachment: None,
+        });
+
+        rpass.set_pipeline(&self.pipeline);
+        rpass.set_vertex_buffers(&[(&self.vertex_buffer, 0)]);
+        rpass.set_index_buffer(&self.index_buffer, 0);
+        self.vertex_count = 0;
+        self.index_count = 0;
+
+        //self.update_uniform_buffer(&matrix)?;
+        let temp_buf = device
+            .create_buffer_mapped(64, wgpu::BufferUsage::TRANSFER_SRC)
+            .fill_from_slice(cast_slice(&matrix));
+
+        let mut copy_encoder =
```

Comment:
You should be able to keep using the original `encoder` here (without submitting it here)